<!DOCTYPE html>
<html lang="en">

  <head> 

    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
ga('create', 'UA-115753061-1', 'auto');

// Replace the following lines with the plugins you want to use.
ga('require', 'eventTracker', {
  attributePrefix: 'data-'
});
ga('require', 'outboundLinkTracker');
ga('require', 'urlChangeTracker');
// ...
ga('send', 'pageview');
    </script>
    <script async src="https://www.google-analytics.com/analytics.js"></script>
    <script async src="autotrack.js"></script>


    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css"> 
    <script src="anchor.js"></script>

  </head>

  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li class="active"><a href="index.html">Home</a></li>
            <li><a href="getting_started.html">Getting Started</a></li>
            <li><a href="input_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>
    </nav>


    <div id="front_page">

      <div class="col-md-10 col-md-offset-1">

        <div class="page-header">
          <h1><img class="relate" src="figures/relate.png"></h1>
          <h4 id="Overview">Software to estimate genome-wide genealogies for thousands of samples</h4>
          <p><b>Relate</b> estimates genome-wide genealogies in the form of trees that adapt to changes in local ancestry caused by recombination. The method, which is scalable to thousands of samples, is described in the following paper. Please cite this paper if you use our software in your study.</p>   
          <p>
          <b>Citations:</b> 
          <ul>
						<li>(original Relate paper) Leo Speidel, Marie Forest, Sinan Shi, Simon Myers.<br>
							A method for estimating genome-wide genealogies for thousands of samples. <a href="https://www.nature.com/articles/s41588-019-0484-x">Nature Genetics 51: 1321-1329, 2019</a>.</b></li>
						<li>(update, v1.1.*) Leo Speidel, Lara Cassidy, Robert W. Davies, Garrett Hellenthal, Pontus Skoglund, Simon R. Myers.<br>
							Inferring population histories for ancient genomes using genome-wide genealogies. <a href="https://academic.oup.com/mbe/article/38/9/3497/6299394">Molecular Biology and Evolution 38: 3497â€“3511, 2021</a>.</b></li>
				</ul>
        <b>Contact:</b> leo.speidel@outlook.com<br/> 
        <b>Website:</b> <a href = "https://leospeidel.com">https://leospeidel.com</a>
          </p>
        </div> 

      </div>



      <div class="container-fluid">
        <div class="col-md-10 col-md-offset-1">

          <!-- 1st row -->
          <div class="row">

            <div class="col-md-8">

              <div class="panel panel-info">
                <div class="panel-body">  
                  <img class="treeplot" src="figures/treeplot.png" >
                </div>
              </div>

            </div>

            <div class="col-md-4">

              <div class="row">
                <h4 id="Download">Download </h4>

                <p><b>Relate</b> is available for academic use. To see rules for non-academic use, please read the <a href="download/20180411_15670_RELATE_Academic_Usage_Licence.pdf" download>LICENCE</a> file, which is included with each software download.</p>

                <h5 id="Binaries">Pre-compiled binaries (last updated: 14/06/2022)</h5> 
                <form role="form" data-toggle="validator" action="download/relate_v1_linux.tar">
                  <div class="form-group">
                    <div class="checkbox">
                      <label>
                        <input type="checkbox" id="terms" oninvalid="this.setCustomValidity('Required')" onchange="this.setCustomValidity('')" required>
                        I agree with the <a href="download/20180411_15670_RELATE_Academic_Usage_Licence.pdf">terms and conditions</a>
                      </label>
                      <div class="help-block with-errors"></div>
                    </div>
                  </div>
                  <div class="form-group">
                    <button type="submit" 
                            class="btn btn-link" 
                            formaction="download/relate_v1.1.9_x86_64_dynamic.tgz"   
                            data-on="click"
                            data-event-category="Relate download (x86_64, dynamic)"
                            data-event-action="click">Linux (x86_64, dynamic) - v1.1.9</button><br/>
                          <button type="submit" 
                                  class="btn btn-link" 
                                  formaction="download/relate_v1.1.9_x86_64_static.tgz"
                                  data-on="click"
                                  data-event-category="Relate download (x86_64, static)"
                                  data-event-action="click">Linux (x86_64, static) - v1.1.9</button><br/>
                                <button type="submit" 
                                        class="btn btn-link" 
                                        formaction="download/relate_v1.1.9_MacOSX_Intel.tgz"
                                        data-on="click"
                                        data-event-category="Relate download (Mac_OSX)"
																				data-event-action="click">Mac OSX (Intel) - v1.1.9</button><br/>     
																<button type="submit" 
																				class="btn btn-link" 
																				formaction="download/relate_v1.1.9_MacOSX_M1.tgz"
																				data-on="click"
																				data-event-category="Relate download (Mac_OSX)"
																				data-event-action="click">Mac OSX (M1) - v1.1.9</button>     
                  </div>
                </form>

								<h5 id="Github repo">Github repository</h5> 
								<p>Alternatively, you can <b>compile your own version</b> by downloading the source code from this
							     <a href="https://github.com/MyersGroup/relate">github repository</a>.
								</p>	

                <p>In the downloaded directory, we have included a toy data set. You can try out <b>Relate</b> using this toy data set by following the instructions on our <a href="getting_started.html">getting started</a> page.</p>

                <p>If you have any problems getting the program to work on your machine or would like to request an executable for a platform not shown here, please send a message to leo.speidel [at] outlook [dot] com. </p>

                <p>We document changes to previous versions in a <a href="#ChangeLog">change-log</a>.</p>
              </div>

            </div>

          </div>

          <!-- 2nd row -->
          <div class="row">
            <h3 id="news">What's new?</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">
            <div class="col-md-12">
							Versions >1.1.0 include a few new features such as:
              <ul>
                <li>Allowing for non-contemporary samples, e.g., ancient DNA: Simply add the --sample_ages option.</li>
                <li>Allowing for haploid organisms: set the ID_2 column in the sample file to NA and Relate will work for haploid organisms.</li>
                <li>Improved coalescence rate estimation: The algorithm has been improved and behaviour has changed slightly. The output now always contains all trees, and the threshold parameter only affects iterations leading up to the last one. The threshold parameter now is a float in [0,1], specifying the proportion of trees to throw away (Default = 0.5).</li>
                <li>Arbitrary strings as chromosome IDs: in scripts such as EstimatePopulationSize.sh, use the --chr option to specify a file containing chromosome name.</li>
                <li>More flexible epochs for coalescence rate estimation: use the option --bins to specify epoch boundaries in the format 10^seq(x,y,stepsize).
              </ul>
            </div>
            <div class="col-md-8 col-md-offset-2">
              <img src="figures/Relate_workflow.png">
            </div>
          </div>

          <!-- 2nd row -->
          <div class="row">
            <h3 id="1000GP">1000 Genomes coalescence rates, allele ages, and selection p-values</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">
            <div class="col-md-12">
              <p>We deposited Relate-estimated coalescence rates, allele ages, and p-values for evidence of positive selection for the 1000 Genomes Project <a href="https://zenodo.org/record/3234689">here</a>.<br/>
              These were obtained by estimating the joint genealogy of all 1000 GP populations and then extracted the embedded genealogy for each population.
              For the genealogy of each population, we jointly estimated the population size history and branch lengths. 
              Variants segregating in more than one population therefore have correlated but different allele ages in each population.</p>
            </div>
          </div>

          <!-- 2nd row -->
          <div class="row">
            <h3 id="relater">Relater: R package for handling Relate output files</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">
            <div class="col-md-12">
              <p>I made an <a href="https://github.com/leospeidel/relater">R package</a> for handling Relate output files (MIT license). It's still under development but the main functionality is working.</p>
              You can install this package, e.g., using
              <pre><code>library(devtools)
devtools::install_github("leospeidel/relater")</code></pre>
            </div>
          </div>

          <!-- 2nd row -->
          <div class="row">
            <h3 id="relater">relate_lib: C++ library for parsing Relate files and tree sequence conversion</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">
            <div class="col-md-12">
              <p>I made a <a href="https://github.com/leospeidel/relate_lib">C++ library</a> for handling Relate output files (MIT license). It's still under development but the main functionality is working.</p>
              <p> This contains a binary for conversion of tree sequence format to anc/mut format and back. (Note that conversions are not loss-less).</p>
              <pre><code>relate_lib/bin/Convert --mode ConvertFromTreeSequence \
              --anc example.anc.gz \
              --mut example.mut.gz \
              -i example</code></pre>
              <pre><code>relate_lib/bin/Convert --mode ConvertToTreeSequence \
              --anc example.anc.gz \
              --mut example.mut.gz \
              -o example</code></pre>
            </div>
          </div>

          <!-- 2nd row -->
          <div class="row">
            <h3 id="Modules">Add-on modules</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="PreparingInputFiles"><a href="input_data.html">Preparing input files</a></h4></div>
                <div class="panel-body">  
                  The input file format of <b>Relate</b> is the haps/sample file format used as an output format of ShapeIt. We provide code to convert files from the hap/legend/sample and vcf file formats. <b>Relate</b> also comes with tools for determining the ancestral allele using an ancestral genome and tools for filtering SNPs using a genomic mask.    
                </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="ExtractTrees"><a href="modules.html#Extract">Extracting subtrees and trees of interest</a></h4></div>
                <div class="panel-body">  
                  We provide code for converting individual trees of interest into .newick format. This file format is convenient, for instance, when visualising a tree. We also provide code for extracting genealogies corresponding to a subpopulation.
                </div>
              </div>
            </div>

          </div>

          <!-- 4th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="PlotMarginalTrees"><a href="modules.html#TreeView">Plot marginal trees</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/tree.jpg">
                  <b>Relate</b> can be used to plot a marginal tree of interest. This tree corresponds to the LCT region, where a mutation at SNP rs4988235 is believed to be responsible for Lactose tolerance in Europeans (here GBR). We can see that the derived allele at this SNP has spread rapidly in GBR, which is indicative of strong positive selection. </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="EstimatePopulationSize"><a href="modules.html#CoalescenceRate">Estimating population sizes and separation histories</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/population_size.jpg">
                  <b>Relate</b> can be used to estimate population sizes and separation histories between populations. The figure shows the separation history between FIN and GBR in the 1000GP data set. The inset shows the matrix of coalescent rates between pairs of haplotypes 9,000 years ago. Rows and columns are sorted by population labels of haplotypes, as indicated by the colour on the left of the matrix. </div>
              </div>
            </div>

          </div>

          <!-- 5th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="EstimateMutationRates"><a href="modules.html#MutationRate">Estimating mutation rates</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/mutation_rate.jpg">
                  <b>Relate</b> can be used to estimate mutation rates through time. The figure shows TCC to TTC mutation rates for all 26 populations in the 1000GP data set. Trends shared between mutation categories are eliminated by dividing by the overall average mutation rate. For each population, the mutation rate is normalized such that its mean over time equals one. Consistent with previous estimates, we see an increase in the mutation rate of TCC to TTC mutations about 10,000 to 30,000 years ago in Europeans and Southern Asians.</div>
              </div>
            </div>


            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="DetectSelection"><a href="modules.html#Selection">Detecting positive selection</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/selection.jpg">
                  <b>Relate</b> can be used to detect evidence for positive selection. We calculate a p-value for selection evidence that quantifies how quickly a mutation has spread in the population. The figure shows a manhattan plot for GBR which indicates clear peaks in the LCT and MHC regions, which are known targets of positive selection.</div>
              </div>
            </div>

          </div>

          <!-- 6th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="aDNA"><a href="getting_started.html">Infer genealogies for ancient samples</a></h4></div>
                <div class="panel-body"> 
                 <img src="figures/plot_aDNA.png">
                 <b>Relate</b> can infer genealogies for non-contemporary samples, such as high coverage ancient genomes or time-stamped samples of bacteria or viruses. This plot was generated using the TreeViewSamples.sh script, using 100 sampled branch lengths. The tree represents the posterior mean times, and "error bars" indicate the 0.025 and 0.975 quantiles of the posterior density of coalescence ages.
                </div> 
              </div>
            </div>


            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="SampleBranchLengths"><a href="modules.html#SampleBranchLengths">Sample branch lengths</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/plot_clues.png">
                  Using the SampleBranchLengths.sh script, <b>Relate</b> can sample branch lengths from the posterior, which can then be used e.g., by CLUES to infer allele frequency trajectories and selection coefficients using an importance sampling scheme. See the <a href="https://github.com/35ajstern/clues">CLUES</a> repo for more details.
                </div>
              </div>
            </div>

          </div>


        </div>
      </div>


      <div class="container-fluid">

        <!-- 1st row -->
        <div class="row">
          <div class="col-md-10 col-md-offset-1">

            <h3 id="ChangeLog">Change-log</h3>
            <hr/>

            <div class="panel-group" id="accordion2">

							<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse29">
											v1.1.9
										</a>
									</h4>
								</div>
								<div id="collapse29" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 14 June 2022 <br/>     
										<ul>
											<li>Fixed bug in TreeView where singletons were not plotted.</li>
										</ul>
									</div>
								</div>
							</div>


							<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse28">
											v1.1.8
										</a>
									</h4>
								</div>
								<div id="collapse28" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 7 November 2021 <br/>     
										<ul>
											<li>Made ConvertToTreeSequence consistent with relate_lib version.</li>
											<li>Bug fix in SummarizeMutationRates (assumed float instead of double, leading to Segfault).</li>
										</ul>
									</div>
								</div>
							</div>

          		<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse27">
											v1.1.7
										</a>
									</h4>
								</div>
								<div id="collapse27" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 16 June 2021 <br/>     
										<ul>
										  <li>Updated initialisation of bl with ancients.</li>
                      <li>Small bug fixes.</li>
										</ul>
									</div>
								</div>
							</div>

			     		<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse26">
											v1.1.6
										</a>
									</h4>
								</div>
								<div id="collapse26" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 17 Feb 2021 <br/>     
										<ul>
											<li>Added option to exclude transitions in branch length estimation (useful for ancient genomes).</li>
										</ul>
									</div>
								</div>
							</div>

			     		<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse25">
											v1.1.5
										</a>
									</h4>
								</div>
								<div id="collapse25" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 14 Dec 2020 <br/>     
										<ul>
											<li>Modified tree builder when ancients samples are specified (see Colate paper for details).</li>
										</ul>
									</div>
								</div>
							</div>

							<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse24">
											v1.1.4
										</a>
									</h4>
								</div>
								<div id="collapse24" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 30 Oct 2020 <br/>     
										<ul>
											<li>Fixed a in ConvertFromVcf that caused empty haps/sample files to be outputted</li>
											<li>Added a function in RelateExtract to get branch frequencies for every SNP.</li>
										</ul>
									</div>
								</div>
							</div>

							<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse23">
											v1.1.3
										</a>
									</h4>
								</div>
								<div id="collapse23" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 7 Oct 2020 <br/>     
										<ul>
											<li>Fixed a major bug (v1.1.* onwards) in EstimatePopulationSize.sh when setting seed (not affected when seed is not set).</li>
											<li>Added RelateSlurm.sh and RelateLSF.sh</li>
											<li>Added a function to retrospectively map a mutation to trees in RelateExtract</li>
										</ul>
									</div>
								</div>
							</div>

							<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse22">
											v1.1.2
										</a>
									</h4>
								</div>
								<div id="collapse22" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 12 June 2020 <br/>     
										<ul>
											<li>Fixed bug in PrepareInputFiles, when downsampling individuals (introduced in v1.1.0).</li>
											<li>Changed parameters for branch lengths estimation for aDNA, to improve mixing properties of MCMC.</li>
											<li>Changed function for anc to newick conversion, to allow multiple trees to be extracted.</li>
										</ul>
									</div>
								</div>
							</div>

							<div class="panel panel-default">
								<div class="panel-heading">
									<h4 class="panel-title">
										<a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse21">
											v1.1.1
										</a>
									</h4>
								</div>
								<div id="collapse21" class="panel-collapse collapse">
									<div class="panel-body">
										Date: 25 May 2020 <br/>     
										<ul>
											<li>Changed filename used for CLUES and PALM from *.palm to *.timeb.</li>
											<li>Fixed bug in EstimatePopulationSize.sh</li>
										</ul>
									</div>
								</div>
							</div>


              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapse20">
                      v1.1.0
                    </a>
                  </h4>
                </div>
                <div id="collapse20" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 20 May 2020 <br/>     
										<ul>
											<li>Allowing for non-contemporary samples, e.g., ancient DNA: Simply add the --sample_ages option.</li>
											<li>Allowing for haploid organisms: set the ID_2 column in the sample file to NA and Relate will work for haploid organisms.</li>
											<li>Improved coalescence rate estimation: The algorithm has been improved and behaviour has changed slightly. The output now always contains all trees, and the threshold parameter only affects iterations leading up to the last one. The threshold parameter now is a float in [0,1], specifying the proportion of trees to throw away (Default = 0.5).</li>
											<li>Arbitrary strings as chromosome IDs: in scripts such as EstimatePopulationSize.sh, use the --chr option to specify a file containing chromosome name.</li>
											<li>More flexible epochs for coalescence rate estimation: use the option --bins to specify epoch boundaries in the format 10^seq(x,y,stepsize).</li>
										</ul>
                  </div>
                </div>
              </div>

            </div>

            <div class="panel-group" id="accordion_version">

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion_version" href="#collapse19">
                      v1.0.*
                    </a>

										<h5 id="Binaries">Pre-compiled binaries (last updated: 10/01/2020)</h5> 
										<form role="form" data-toggle="validator" action="download/relate_v1_linux.tar">
											<div class="form-group">
												<div class="checkbox">
													<label>
														<input type="checkbox" id="terms" oninvalid="this.setCustomValidity('Required')" onchange="this.setCustomValidity('')" required>
														I agree with the <a href="download/20180411_15670_RELATE_Academic_Usage_Licence.pdf">terms and conditions</a>
													</label>
													<div class="help-block with-errors"></div>
												</div>
											</div>
											<div class="form-group">
												<button type="submit" 
																class="btn btn-link" 
																formaction="download/v1.0.17/relate_v1.0.17_x86_64_dynamic.tgz"   
																data-on="click"
																data-event-category="Relate download (x86_64, dynamic)"
																data-event-action="click">Linux (x86_64, dynamic) - v1.0.17</button><br/>
															<button type="submit" 
																			class="btn btn-link" 
																			formaction="download/v1.0.17/relate_v1.0.17_x86_64_static.tgz"
																			data-on="click"
																			data-event-category="Relate download (x86_64, static)"
																			data-event-action="click">Linux (x86_64, static) - v1.0.17</button><br/>
																		<button type="submit" 
																						class="btn btn-link" 
																						formaction="download/v1.0.17/relate_v1.0.17_MacOSX.tgz"
																						data-on="click"
																						data-event-category="Relate download (Mac_OSX)"
																						data-event-action="click">Mac OSX - v1.0.17</button>                   
											</div>
										</form>

                    <div id="collapse19" class="panel-collapse collapse">
                      <div class="panel-body">

                        <div class="panel-group" id="accordion">


                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse18">
                                  v1.0.17
                                </a>
                              </h4>
                            </div>
                            <div id="collapse18" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 10th January 2020<br/>     
                                <ul>
                                  <li>Input files can now have CHR, RSID, ANC, DER columns of type strings (up to 1024 chars).</li>
                                  <li>Added a scripts named TreeViewMutation.sh, taking same input as TreeView.sh, but different visualisation (bp_of_interest has to be SNP).</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse17">
                                  v1.0.16
                                </a>
                              </h4>
                            </div>
                            <div id="collapse17" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 2nd September 2019<br/>     
                                <ul>
                                  <li>Fixed a bug in ConvertToTreeSequence that meant internal nodes were flagged as sample nodes.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse16">
                                  v1.0.15
                                </a>
                              </h4>
                            </div>
                            <div id="collapse16" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 13th August 2019<br/>     
                                <ul>
                                  <li>Fixed a small bug that meant memory was misallocated when the --memory value was set too large.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse15">
                                  v1.0.14
                                </a>
                              </h4>
                            </div>
                            <div id="collapse15" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 24th July 2019<br/>     
                                <ul>
                                  <li>Fixed a bug in PrepareInputFiles.sh that meant the script terminated before filtering SNPs using a genomic mask when input was not gzipped.</li>
                                  <li>Changed the default for y axis scale in TreeView to not be a log-scale.</li>
                                </ul>
                              </div>
                            </div>
                          </div>


                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse14">
                                  v1.0.13
                                </a>
                              </h4>
                            </div>
                            <div id="collapse14" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 27th May 2019<br/>     
                                <ul>
                                  <li>Modified default criteria to terminate population size estimation. Now terminating when iteration at least 2 and "mean absolute error"/mu less than 0.1.</li>
                                  <li>Modified memory allocation in Relate, where we were allocating too much memory for very small sample sizes (N less than 10).</li>
                                  <li>Modified error message when poplabels file is misspecified in FinalizePopulationSize - you can just rerun this step again at the end.</li>
                                  <li>Included the script RelateSGE.sh for running Relate on a SGE cluster.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse13">
                                  v1.0.12
                                </a>
                              </h4>
                            </div>
                            <div id="collapse13" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 16th May 2019<br/>     
                                <ul>
                                  <li>Fixed a bug in FinalizePopulationSize --mode EstimatePopulationSize, where I had forgotten to close a file leading to an error on some platforms.</li>
                                  <li>Allowing .poplabels file to be tab separated.</li>
                                  <li>Fixed a bug in TreeView which threw an error when completely fixed SNPs were included in the haps/sample file.</li>
                                </ul>
                              </div>
                            </div>
                          </div>



                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse12">
                                  v1.0.11
                                </a>
                              </h4>
                            </div>
                            <div id="collapse12" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 14th May 2019<br/>     
                                <ul>
                                  <li>Fixed a bug in FinalizePopulationSize introduced in v1.0.9, which threw an error message for correct input files.</li>
                                  <li>Fixed a bug in SampleBranchLengths which was introducing small rounding errors.</li>
                                </ul>
                              </div>
                            </div>
                          </div>



                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse11">
                                  v1.0.10
                                </a>
                              </h4>
                            </div>
                            <div id="collapse11" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 13th May 2019<br/>     
                                <ul>
                                  <li>Fixed a bug in SampleBranchLengths.sh, where a filename was hard-coded by mistake.</li>
                                </ul>
                              </div>
                            </div>
                          </div>


                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse10">
                                  v1.0.9
                                </a>
                              </h4>
                            </div>
                            <div id="collapse10" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 3rd May 2019<br/>     
                                <ul>
                                  <li>Implemented a module which samples branch lengths from the posterior given tree topologies and effective population size histories.</li>
                                  <li>Added some more detail to documentation for RelateSelection.</li>
                                  <li>Added a more meaningful error message when .poplabels file is misspecified in EstimatePopulationSize.sh.</li>
                                </ul>
                              </div>
                            </div>
                          </div>


                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse9">
                                  v1.0.8
                                </a>
                              </h4>
                            </div>
                            <div id="collapse9" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 15th March 2019<br/>     
                                <ul>
                                  <li>Now requiring different input and output names for DetectSelection.sh when it overwrites files otherwise.</li>
                                  <li>DetectSelection.sh now works also if additional columns are not appended to the .mut file using ./RelateFileFormats --mode GenerateSNPAnnotations. However, code is more efficient if these columns are appended.</li>
                                  <li>Switched to using R package cowplot in TreeView.sh.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse8">
                                  v1.0.7
                                </a>
                              </h4>
                            </div>
                            <div id="collapse8" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 25th Feburary 2019<br/>     
                                <ul>
                                  <li>Fixed bug in the painting which affects tree topologies sometimes, but has only a small effect (not visible in distance measures when comparing to truth in simulations).</li> 
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse7">
                                  v1.0.6
                                </a>
                              </h4>
                            </div>
                            <div id="collapse7" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 6th Feburary 2019<br/>     
                                <ul>
                                  <li>I added a function for converting anc/mut output files to tree sequence file format (tskit). Currenty, some information is lost by this conversion.</li> 
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse6">
                                  v1.0.5
                                </a>
                              </h4>
                            </div>
                            <div id="collapse6" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 18th October 2018<br/>     
                                <ul>
                                  <li>I had introduced a bug in update v1.0.4 in the function RelateExtract --mode TreeAtSNPAsNewick which has been fixed.</li> 
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse5">
                                  v1.0.4
                                </a>
                              </h4>
                            </div>
                            <div id="collapse5" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 13th October 2018<br/>     
                                <ul>
                                  <li>Substituted zcat with gunzip -c to fix a bug in Macs.</li>
                                  <li>Added requirement of R version >= 3.3.1 for TreeView because of a known bug with grid.draw() in older versions.</li>
                                  <li>Fixed bug in RelateExtract --mode TreeAtSNPAsNewick, which outputted the first tree when snp_of_interest was not a SNP in the data set. Changed option snp_of_interest to pos_of_interest and this function now prints the tree at the position of interest. In addition, output filename was not using the option -o, which has been corrected.</li> 
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse4">
                                  v1.0.3
                                </a>
                              </h4>
                            </div>
                            <div id="collapse4" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 30th August 2018<br/>     
                                <ul>
                                  <li>Implemented pipeline for calculating p-values for selection evidence. Updated corresponing entry in the documentation.</li>
                                  <li>Bug fix in RelateSelection --mode Frequency: Previous version had a bug whenever two internal nodes had same age.</li>
                                  <li>Fixed function RelateFileFormats --mode ConvertFromVcf which previously parsed the vcf incorrectly in some cases.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse3">
                                  v1.0.2
                                </a>
                              </h4>
                            </div>
                            <div id="collapse3" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 16th July 2018<br/>     
                                <ul>
                                  <li>Implemented parallelization of module EstimatePopulationSize.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse2">
                                  v1.0.1
                                </a>
                              </h4>
                            </div>
                            <div id="collapse2" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 30th June 2018<br/>
                                <ul>
                                  <li>Bug fix in parsing function of haps/sample.</li>
                                </ul>
                              </div>
                            </div>
                          </div>

                          <div class="panel panel-default">
                            <div class="panel-heading">
                              <h4 class="panel-title">
                                <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse1">
                                  v1.0.0
                                </a>
                              </h4>
                            </div>
                            <div id="collapse1" class="panel-collapse collapse">
                              <div class="panel-body">
                                Date: 4th June 2018
                                <ul>
                                  <li> Initial release.</li>
                                </ul>
                              </div>
                            </div>
                          </div>
                        </div>

                      </div>

                    </div>
                </div>
              </div>
            </div>
          </div>

        </div>
      </div>

    </div>

    <script>
      anchors.add();  
    </script>


  </body>
</html>        

