<!DOCTYPE html>
<html lang="en">


  <head>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-115753061-1"></script>
    <script>
window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());

gtag('config', 'UA-115753061-1');
    </script>


    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css">
    <script src="anchor.js"></script>

  </head>



  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li><a href="index.html">Home</a></li>
            <li><a href="getting_started.html">Getting Started</a></li>
            <li class="active"><a href="input_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>

    </nav>


    <!-- CONTENT -->


    <div class="wrapper">

      <!-- Sidebar Holder -->
      <nav id="sidebar">

        <div class="sidebar-header">
          <h3>Overview</h3>
        </div>

        <ul class="list-unstyled components">
          <li>
            <a href="#FileFormat">Input file formats</a>
          </li>
          <li>
            <a href="#ConvertToFileFormat">Convert to haps/sample file format</a>
          </li> 
          <li>
            <a href="#Prepare">Preparing input files</a>
          </li> 
          <li>
            <a href="#Convert">Convert output to other formats</a>
          </li> 
          <li>
            <a href="#Data">External resources</a>
          </li> 
        </ul>

      </nav>

      <!-- Page Content Holder -->
      <div id="content">

        <div class="container-fluid">

          <!-- row -->
          <div class="row">
            <div class="col-md-10 col-md-offset-1">


              <h3 id="FileFormat">Input data file format</h3>

              <p>The input file format for <b>Relate</b> is the haps/sample file format, which is used as an output file format by SHAPEIT2 (see <a href="http://www.shapeit.fr/pages/m02_formats/hapssample.html">here</a>).</p>

              <table class="table">
                <tbody>
                  <tr>
                    <a class="anchor" name=".haps"> </a>
                    <td>.haps</td>
                    <td>The first five columns are:
                      <ol>
                        <li>Chromosome number [integer]</li>
                        <li>SNP ID [string]</li>
                        <li>SNP position [integer]</li>
                        <li>Ancestral allele [char]</li>
                        <li>Alternative allele [char]</li>
                      </ol>
                      Then, the successive column pairs correspond to the haplotypes carried by an individual at that SNP.<br/>
                      <b>Please make sure that the ancestral allele is represented by 0 and the alternative allele is represented by 1.</b>
                      <pre><code>7 SNP1 123 A G 0 0 1 0 0 0 1 1<br/>7 SNP2 456 T C 0 1 1 0 0 1 0 1<br/>7 SNP3 789 A T 0 1 1 0 1 1 1 1</code></pre>
                      This file is space delimited.
                    </td>
                  </tr>
                  <tr>
                    <a class="anchor" name=".sample"></a>
                    <td>.sample</td>
                    <td>The three columns are:
                      <ol>
                        <li>First individual ID</li>
                        <li>Second individual ID</li>
                        <li>Missing data proportion</li>
                      </ol>
											<b>Diploid organisms:</b>
                      <pre><code>ID_1 ID_2 missing<br/>0    0    0<br/>UNR1 UNR1 0<br/>UNR2 UNR2 0<br/>UNR3 UNR3 0<br/>UNR4 UNR4 0<br/></code></pre>
											<b>Haploid organisms:</b>
											<pre><code>ID_1 ID_2 missing<br/>0    0    0<br/>UNR1 NA 0<br/>UNR2 NA 0<br/>UNR3 NA 0<br/>UNR4 NA 0<br/></code></pre>
											This file is space delimited.<br/>
                    </td>
                  </tr>
                </tbody>
              </table>

              <p>Recombination rates are specified using a genomic map.</p>
              <table class="table">
                <tbody>
                  <tr>
                    <a class="anchor" name=".map"></a>
                    <td>.map</td>
                    <td>This is the standard file format for genetic recombination maps. The three columns are:
                      <ol>
                        <li>Position (b) [integer]</li>
                        <li>Recombination rate (cM/Mb) [float]</li>
                        <li>Genetic position (cM) [float]</li>
                      </ol>
                      Denoting the ith entry of the three columns by p[i], r[i], rdist[i], the following equation holds <br/> 
                      r[i] = (rdist[i+1] - rdist[i])/(p[i+1] - p[i]) * 1e6 

                      <pre><code>pos COMBINED_rate Genetic_Map<br/>0        2.8074 0.4103<br/>2529     2.7778 0.4174<br/>2601     2.9813 0.4176</code></pre>
                      This file is space delimited.
                    </td>
                  </tr>
                </tbody>
              </table>

              <p>For some add-on modules of <b>Relate</b>, we also want to specify population labels. The .poplabels file format is identical to the sample file in the <a href="http://www.shapeit.fr/pages/m02_formats/haplegsample.html">hap/legend/sample file format</a>.</p>
              <table class="table">
                <tbody>
                  <tr>
                    <a class="anchor" name=".poplabels"></a>
                    <td>.poplabels</td>
                    <td>The four columns are:
                      <ol>
                        <li>Individual ID as specified in the .sample file [string]</li>
                        <li>Population label [string]</li>
                        <li>Group label [string]</li>
                        <li>Sex [integer]</li>
                      </ol>
											<b>Diploid organisms:</b>
                      <pre><code>sample population group sex<br/>UNR1 PJB SAS NA<br/>UNR2 JPT EAS NA<br/>UNR3 GBR EUR NA<br/>UNR4 YRI AFR NA</code></pre>
											<b>Haploid organisms:</b>
											<pre><code>sample population group sex<br/>UNR1 PJB SAS 1<br/>UNR2 JPT EAS 1<br/>UNR3 GBR EUR 1<br/>UNR4 YRI AFR 1</code></pre>
											For haploid organisms (or haplotype level annotation in diploid organisms) set the SEX column to 1.<br/>
                      This file is space delimited.<br>
											<b>Samples must be listed in the same order as in the .sample file above.</b><br/>
                    </td>
                  </tr>
                </tbody>
              </table>




            </div>
          </div>

          <!-- row -->
          <div class="row">
            <div class="col-md-10 col-md-offset-1">

              <h3 id="ConvertToHapsSample">Convert to haps/sample file format</h3>
              <hr/>

            </div>
          </div>



          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">

              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="ConvertFromHapLegendSample">Convert from hap/legend/sample</h4></div>
                <div class="panel-body">  
                  Converts from the hap/legend/sample file format (used for reference panels in e.g., IMPUTE2, see <a href="http://www.shapeit.fr/pages/m02_formats/haplegsample.html">here</a>) to the haps/sample file format. The code deletes non-biallelic SNPs.

                  <dl class="dl-horizontal">
                    <dt>--haps</dt>
                    <dd>Filename of .haps file.</dd>
                    <dt>--sample</dt>
                    <dd>Filename of .sample file</dd>
                    <dt>-i,--input</dt>
                    <dd>Filename (without file extension) of the hap/legend/sample files. The input files can be gzipped; we look for files with extensions hap,hap.gz/legend,legend.gz/sample,sample.gz, with preference given to the non-gzipped file.</dd>
                    <dt>--chr</dt>
                    <dd>Optional: specifies a chromosome index [int] (used as the first column in the haps file). Default: 0.</dd>
                  </dl>

                  <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode ConvertFromHapLegendSample \
                 --haps example.haps \
                 --sample example.sample \
                 -i example \
                 --chr 1  </code></pre>
                  Output: example.haps, example.sample.

                </div>
              </div>
            </div>
          </div>

          <div class="row">

            <div class="col-md-10 col-md-offset-1">

              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="ConvertFromVcf">Convert from vcf</h4></div>
                <div class="panel-body">  

                  Converts from the vcf file format to the haps/sample file format. During the process non-biallelic SNPs are dropped.

                  <dl class="dl-horizontal">
                    <dt>--haps</dt>
                    <dd>Filename of .haps file.</dd>
                    <dt>--sample</dt>
                    <dd>Filename of .sample file</dd>
                    <dt>-i,--input</dt>
                    <dd>Filename (without file extension) of the vcf file. The input files can be gzipped; we look for files with extensions hap,hap.gz/legend,legend.gz/sample,sample.gz, with preference given to the non-gzipped file.</dd>
                    <dt>--chr</dt>
                    <dd>Optional: specifies a chromosome index [int] (used as the first column in the haps file). Default: 0.</dd>
                  </dl>

                  <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode ConvertFromVcf \
                 --haps example.haps \
                 --sample example.sample \
                 -i example </code></pre>
                  Output: example.haps, example.sample.

                </div>
              </div>
            </div>

            <div class="row">

              <div class="col-md-10 col-md-offset-1">

                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="ConvertFromMs">Convert from ms output format</h4></div>
                  <div class="panel-body">  

                    R script that converts from the output file format used by ms to the haps/sample file format.<br/>
                    The script removes any sites with multiple mutations.<br/>
                    If the input file contains more than one simulation, these get separated into "chromosomes".<br/><br/>

                    <a href = "./download/ms2haps.R">Download the script.</a><br><br/>

                    <pre><code> Rscript ms2haps.R infile.ms example nsites</code></pre>

                    <dl class="dl-horizontal">
                      <dt>infile.ms</dt>
                      <dd>Input filename with file extension.</dd>
                      <dt>example</dt>
                      <dd>Output filename withouth file extension.</dd>
                      <dt>nsites</dt>
                      <dd>(Optional) Number of simulated sites. Default value is 1. This is multiplied to the positions.</dd>
                    </dl>
                    Output: example.haps, example.sample.

                  </div>
                </div>
              </div>


            </div>

            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">

                <h3 id="Prepare">Preparing input files</h3>
                <hr/>
                <p>The following code can be used to prepare the input data. We recommend to apply all steps (if applicable) in the listed order. </p>
                <p>All input files can be gzipped.</p>


                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="PrepareInput">Prepate input files</h4></div>
                  <div class="panel-body">  

                    Prepares input files for <b>Relate</b>. The script will remove non-biallelic SNPs, flip haplotypes according to the ancestral genome such that the ancestral allele is always denoted by 0, remove samples if specified, filter SNPs and adjust distances between SNPs using a genomic mask if specified, and generate SNP annotations needed for some add-on modules.<br/> 
                    See <a href="input_data.html#Data">here</a> for links to external resources, such as an ancestral genome.

                    <dl class="dl-horizontal">
                      <dt>--haps</dt>
                      <dd>Filename of .haps file.</dd>
                      <dt>--sample</dt>
                      <dd>Filename of .sample file</dd>
                      <dt>-o,--output</dt>
                      <dd>Filename of output files without file extensions. If same as input filenames, input files will be overwritten.</dd>
                      <dt>--ancestor</dt>
                      <dd>Fasta file containing ancestral genome. This is case insensitive.</dd>
                      <dt>--mask</dt>
                      <dd>Optional: Fasta file of same length as the ancestral genome containing a genomic mask. Loci passing the mask are denoted by P, loci not passing the mask are denoted by N. This is case insensitive.</dd>
                      <dt>--remove_ids</dt>
                      <dd>Optional: File containing ids of samples that will be removed from the haps/sample files. One id per line.</dd>
                      <dt>--poplabels</dt>
                      <dd>Optional: File containing population labels of samples.</dd>
                    </dl>

                    <pre><code> PATH_TO_RELATE/scripts/PrepareInputFiles/PrepareInputFiles.sh \
                 --haps example.haps \
                 --sample example.sample \
                 --ancestor ancestor.fa \
                 --mask mask.fa \
                 --remove_ids remove_ids.txt \
                 --poplabels example.poplabels
                 -o example_input </code></pre>
                    Output: example_input.haps, example_input.sample, example_input.dist (if --mask if specified), example_input.poplabels (if remove_ids and poplabels is specified), example_input.annot (if poplabels is specified).

                  </div>
                </div>
              </div>

            </div>


            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">

                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="RemoveNonBiallelicSNPs">Remove non-biallelic SNPs</h4></div>
                  <div class="panel-body">  

                    Removes SNPs that are not biallelic in the data set.
                    <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode RemoveNonBiallelicSNPs \
                 --haps example.haps \
                 -o example_biallelic </code></pre>
                    Output: example_biallelic.haps.

                  </div>
                </div>
              </div>

            </div>


            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">
                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="FlipHapsUsingAncestor">Determine ancestral allele and flip if necessary</h4></div>
                  <div class="panel-body">  

                    The ancestral genome should be saved as a fasta file and aligned to the data set. The code is insensitive to case in the fasta file and deletes SNPs at which neither allele matches the ancestral allele.
                    <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode FlipHapsUsingAncestor \
                 --haps example.haps \
                 --sample example.sample \
                 --ancestor ancestor.fa \
                 -o example_ancestral </code></pre>
                    Output: example_ancestral.haps

                  </div>
                </div>
              </div>
            </div>

            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">
                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="RemoveSamples">Remove samples from data set</h4></div>
                  <div class="panel-body">  

                    Write ids of individuals as specified in the .sample file into remove_ids.txt (one id per line). These indivduals will be removed from the data set. 
                    Argument --poplabels is optional. If specified, it will generate a new poplabels file for the subsetted samples.
                    <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode RemoveSamples \
                 --haps example.haps \
                 --sample example.sample \
                 --poplabels example.poplabels \
                 -i remove_ids.txt \
                 -o example_subset </code></pre>
                    Output: example_subset.haps, example_subset.sample, example_subset.poplabels (if --poplabels is specified).

                  </div>
                </div>
              </div>
            </div>

            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">
                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="FilterHapsUsingMask">Filter SNPs using a genomic mask</h4></div>
                  <div class="panel-body">  

                    The genomic mask needs to be specified as a fasta file of same length as the ancestral genome, such that the xth entry is the xth base of the genome. The passing state should be denoted as P. The code is insensitive to case in the fasta file and deletes all SNPs that are not passing. It also outputs a .dist file that contains the distances between SNPs, adjusted for regions that are not passing in the mask.
                    <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode FilterHapsUsingMask\
                 --haps example.haps \
                 --sample example.sample \
                 --mask genomic_mask.fa \
                 -o example_mask </code></pre>
                    Output: example_mask.haps, example_mask.dist.

                  </div>
                </div>
              </div>
            </div>



            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">
                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="GenerateSNPAnnotations">Generate SNP annotations</h4></div>
                  <div class="panel-body">  

                    This code uses the ancestral genome and population labels of samples to append additional information to the output of <b>Relate</b>. Fileformat for population labels is described <a href="input_data.html#.poplabels">here</a>. The ancestral genome should be saved in fasta file format and aligned to the data set. Argument --ancestor is optional. If argument --mut is specified, output is directly appended to the .mut file.</br>
                    <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode GenerateSNPAnnotations \
                 --haps example.haps \
                 --sample example.sample \
                 --poplabels example.poplabels \
                 --ancestor ancestor.fa \
                 -o example </code></pre>
                    Output: example.annot. If --mut is specified, columns are appended to example.mut.

                  </div>
                </div>
              </div>
            </div>


            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">

                <h3 id="Convert">Convert output to other formats</h3>
                <hr/>
                <p>The following code can be used to convert the output to other file format<p>

              </div>

            </div>

            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">
                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="TreeSequence">Convert to tree sequence file format (<a href="https://tskit.readthedocs.io/en/stable/">tskit</a>)</h4></div>
                  <div class="panel-body">  

                    <p>This function converts anc/mut files inferred by <b>Relate</b> into the tree sequence file format used by <a href="https://tskit.readthedocs.io/en/stable/">tskit</a>. In the current implementation, each tree is stored with new nodes in the tree sequence file format, leading to no compression. In addition, information about how long branches persist, and how many mutations map to a branch are lost by this conversion.
                    </p>

                    <dl class="dl-horizontal">
                      <dt>-i,--input</dt>
                      <dd>Filename of .anc and .mut files without file extension.</dd>
                      <dt>-o,--output</dt>
                      <dd>Filename of the output tree sequence file without file extensions.</dd>
                    </dl>

                    <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode ConvertToTreeSequence \
                 -i example \
                 -o example
                    </code></pre>
                    Output: example.trees.

                  </div>
                </div>
              </div>
            </div>

            <!-- row -->
            <div class="row">

              <div class="col-md-10 col-md-offset-1">
                <div class="panel panel-info">
                  <div class="panel-heading"> <h4 id="TreeAtSNPAsNewick">Extract tree at a SNP of interest in newick format</h4></div>
                  <div class="panel-body">  

                    <p>Extracts a tree at a SNP of interest in newick format.</p>

                    <dl class="dl-horizontal">
                      <dt>--anc</dt>
                      <dd>Filename of the .anc file.</dd>
                      <dt>--mut</dt>
                      <dd>Filename of the .mut file.</dd>
                      <dt>--bp_of_interest</dt>
                      <dd>Integer specfifying the BP position of interest</dd>
                      <dt>-o,--output</dt>
                      <dd>File of the output files without file extensions.</dd>
                    </dl>

                    <pre><code> PATH_TO_RELATE/bin/RelateExtract\
                 --mode TreeAtSNPAsNewick \
                 --anc example.anc \
                 --mut example.mut \
                 --bp_of_interest 3000000 \
                 -o example </code></pre>
                    Output: example_at_.newick.

                  </div>
                </div>
              </div>

            </div>



            <div class="row">
              <div class="col-md-10 col-md-offset-1">

                <h3 id="Data">External resources</h3>
                <hr/>

                <p>Below is a list of external resources that can be useful for Relate-based analyses. These resources were compiled by thrid parties, and their respective conditions (such as citations of corresponding publications) for using these resources apply.</p>

                <h4 id="Humans">Humans</h4>

								Here is a dropbox link to hg37 and hg38 <a href="https://www.dropbox.com/scl/fo/okg8jvlhvbra17m60h2n6/h?rlkey=4he3olpkt0w9db3svsy47b2cr&dl=0"> ancestral genomes, recombination maps, and mappability masks</a>.<br/><br/>

								External links:
                <ul class="list-unstyled">
                  <li><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/supporting/ancestral_alignments/">Ancestral genome (GRCh37)</a></li>
                  <li><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/">Genome mask for 1000GP_Phase3 (GRCh37)</a></li>
                  <li><a href="https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.html">1000GP_Phase3 haplotypes</a></li>
                </ul>
              </div>
            </div>


          </div>

        </div>

        <script>
anchors.add();  
        </script>

  </body>
</html>     
